tests/test_gpu_examples.py

```diff
@@ -1435,6 +1438,195 @@ def test_offload_merge(self):
+    model = self.get_model()
+    output_base = model(data)[0]
+
+    config = LoraConfig(init_lora_weights="pissa", target_modules=["linear"], r=8)
+    peft_model = get_peft_model(deepcopy(model), config)
+    # save the initial model
+    peft_model.peft_config["default"].init_lora_weights = True
+    peft_model.save_pretrained(tmp_path / "init-model")
+    peft_model = peft_model.unload()
+    torch.save(peft_model.state_dict(), tmp_path / "residual-model")
+    del peft_model
+
+    # create 4bit base model
+    base_model = deepcopy(model)
+    base_model.load_state_dict(torch.load(tmp_path / "residual-model"))
+    # sanity check: the base model weights were indeed changed
+    tol = 1e-06
+    assert not torch.allclose(model.linear.weight, base_model.linear.weight, atol=tol, rtol=tol)
+    # quantize the linear layer
+    linear4bit = bnb.nn.Linear4bit(base_model.linear.in_features, base_model.linear.out_features)
+    linear4bit.load_state_dict(base_model.linear.state_dict())
+    linear4bit.to(0)
+    base_model.linear = linear4bit
+    peft_model = PeftModel.from_pretrained(deepcopy(base_model), tmp_path / "init-model")
+    output_quantized_pissa = peft_model(data)[0]
+    # sanity check
+    tol = 1e-06
+    assert not torch.allclose(output_base, output_quantized_pissa, atol=tol, rtol=tol)
+
+    # modify the weights, or else the adapter performs an identity transformation
+    peft_model.base_model.linear.lora_B["default"].weight.data *= 2.0
+    output_finetuned_pissa = peft_model(data)[0]
+    # sanity check
+    tol = 1e-06
+    assert not torch.allclose(output_quantized_pissa, output_finetuned_pissa, atol=tol, rtol=tol)
+
+    # save the model normally
+    peft_model.save_pretrained(tmp_path / "pissa-model")
+    model_loaded = PeftModel.from_pretrained(deepcopy(base_model), tmp_path / "pissa-model")
+    output_loaded = model_loaded(data)[0]
+
+    assert torch.allclose(output_finetuned_pissa, output_loaded, atol=tol, rtol=tol)
+    # sanity check: ranks should still be 8 as initially
+    assert model_loaded.peft_config["default"].r == 8
+    assert model_loaded.base_model.model.linear.lora_A["default"].weight.shape[0] == 8
+
+    # save the model with conversion
+    peft_model.save_pretrained(tmp_path / "pissa-model-converted", convert_pissa_to_lora=tmp_path / "init-model")
+    model_converted = PeftModel.from_pretrained(deepcopy(model), tmp_path / "pissa-model-converted")
+    output_converted = model_converted(data)[0]
+
+    # rank should be double of what it was initially
+    assert model_converted.peft_config["default"].r == 16
+    assert model_converted.base_model.model.linear.lora_A["default"].weight.shape[0] == 16
+    # base model weights should be the same as the initial model
+    assert torch.allclose(
+        model.linear.weight, model_converted.base_model.model.linear.base_layer.weight, atol=tol, rtol=tol
+    )
+    assert torch.allclose(output_finetuned_pissa, output_converted, atol=tol, rtol=tol)
```

Comment:
Instead of marking the whole test as `xfail`, let's remove that decorator. Instead, for this last assert, let's change it to `assert not torch.allclose(output_finetuned_pissa, output_converted, atol=tol, rtol=tol)` with a comment like "# This check is expected to fail when using bnb".

The problem with marking the whole test as `xfail` is that if it fails another, unexpected reason, we don't see it.